### ISSUE TITLE
Public forms always reloads the page on save

### ISSUE THREAD

**Issue by eXistenZe** [2sic/2sxc#2587]
**I'm submitting a ...**
[x] feature request
[x] support request

**...about**  
[x] Content Types or data management

**Current behavior**
If you use public forms to invoke the 2sxc GUI do edit or create data, when you save, the page always reloads.
Ajax reload  avoids the reload but generates a permission error since the users are not admins.

**Expected behavior** 
There should be an option for the save function not to reload the page.
For example when using a js tree to view the data, I need to keep the tree open and this is impossible to do so with automatic reload. Custom forms and APIs seem to be the only alternative.

**Comment by iJungleboy** [comment]
You're in luck - there is a feature we implemented ca. 2 years ago which does this - but it's not documented - sorry :(. 

Here's how it works

1. When the toolbar is created, it will fire an event called `toolbar-init`
2. When fired, the event caries with it a workflow-manager, where you can register custom workflow steps to be executed - like before the edit-dialog opens, after it closes, before refresh etc.
3. These workflow-step can simply return false, to prevent an action from happening

I'll publish it in our Tutorials within the next few days https://2sxc.org/dnn-tutorials/en/razor - but here's a working example just for you ;)

```html
<div class="alert alert-primary" style="width: 50%;">
  <div id='tagWithToolbar1' @Edit.TagToolbar(toolbar: new [] { "toolbar=empty", "+new?contentType=UiEmptyHelloWorld" })>
    Title: @Content.EntityTitle
  </div>
</div>

<script>
  // Attach event-listener to the parent, so we can register the workflow when the toolbar is created
  var parent = document.getElementById('tagWithToolbar1');
  parent.addEventListener('toolbar-init', (e) => {
    console.log("Workflow Demo: Toolbar was initialized - event kicked in - will now register");
    const workflow = e.detail.workflow;

    // This is the workflow step we will register
    var workflowToDisableRefresh = {
      command: 'refresh',   // The command name it's for
      phase: 'before',      // The workflow-step should run before the command is executed
      code: (wfArgs) => {   // The code which should be run
        console.log('Toolbar asked to refresh, will return false to stop it. These are the arguments we got.', wfArgs);
        return false;       // Return false to stop this command from happening
      }
    };

    workflow.add(workflowToDisableRefresh);
  });
</script>
```

**Comment by iJungleboy** [comment]
Full demo here :)

**Comment by eXistenZe** [comment]
I can see how the example works for toolbars, but not for public forms with  $2sxc.cms.run

**Comment by iJungleboy** [comment]
I now created docs https://docs.2sxc.org/js-code/toolbars/workflows.html

I haven't checked the `cms.run` for a long time, but I believe it returns a promise which you can cancel. Please try and write how it worked. I would appreciate a working example so I can add it to the tutorial. 

If that promise-cancel doesn't work, we'll fix it.

**Comment by eXistenZe** [comment]
I believe I copied the examples from here:
https://2sxc.org/en/blog/post/recipe-create-public-forms-with-2sxc
https://2sxc.org/en/apps/app/tutorial-public-forms-with-2sxc-9-30

**Comment by iJungleboy** [comment]
I see that in old times, return false did the trick, but not any more. 

I'll try to get a better implementation using workflow ready for the next release.

**Comment by eXistenZe** [comment]
Don't let it go into top priorities over this.
Events on open or close form are very interesting, but I found a workaround
for my usecase.

When I browse the jstree I set an event that creates a cookie with the ID
of the selected node, and on page load another event that opens the tree up
to the last id'ed node. That way the tree is always on the last node we
selected :) Learned a new trick. Who said cookies are evil? I wonder if
2sxc razor views can get the values of the site cookies as well... another
test for another time.

Stoping the reload would still be interesting cause the modals would still
be opened with the info that the user just created.

Thanks.
